File: problems/0010_regular-expression-matching/backtrace/solution.c
bool isMatch(char * s, char * p) {
    const char p0 = p[0], s0 = s[0];
    const bool sIsEmpty = (s0 == '\0'),
               pIsEmpty = (p0 == '\0');
    if (pIsEmpty)
        return sIsEmpty;
    const bool matchP0 = (!sIsEmpty &&
            (p0 == s0 || p0 == '.'));
    if (p[1] != '\0' && p[1] == '*')
        return (isMatch(s, p + 2) ||
                (matchP0 && isMatch(s + 1, p)));
    else
        return (matchP0 && isMatch(s + 1, p + 1));
}

File: problems/0010_regular-expression-matching/backtrace/solution.js
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function (s, p) {
    'use strict'
    if (p.length == 0)
        return (s.length == 0)
    const p0 = p[0],
          matchP0 = (s.length > 0 &&
            (p0 == s[0] || p0 == '.'))
    if (p.length >= 2 && p[1] == '*')
        return (isMatch(s, p.substring(2)) ||
                (matchP0 && isMatch(s.substring(1), p)))
    else
        return (matchP0 && isMatch(s.substring(1), p.substring(1)))
};
